HACKING.md

```diff
@@ -74,3 +74,79 @@ your implementation's repository: you can set it with
 
 This is harder than writing a new test, but ultimately far more
 useful, as it expands the coverage of every new test.
+
+To add a new runner, you'll need to create a new subclass of Runner, that
+fills in the Runner API. You can find a good skeleton for a new runner in
+`lnprototest/dummyrunner.py`
+
+A completed c-lightning example runner can be found in `lnprototest/clightning/clightning.py`
+
+Here's a short outline of the current expected methods for a Runner.
+
+- `get_keyset`: returns the node's KeySet (`revocation_base_secret`, `payment_base_secret`, `htlc_base_secret`, and `shachain_seed`)
+- `get_node_privkey`: Private key of the node. Used to generate the node id and establish a communication channel with the node under test.
+- `get_node_bitcoinkey`: Private key of the node under test's funding pubkey
+- `has_option`: Returns `None` if node does not support provided option, otherwise 'even' or 'odd' (required or supported). Example input: `option_data_loss_protect` or `option_anchor_outputs`
+- `start`: Starts up / initializes the node under test.
+- `stop`: Stops the node under test and closes the connection.
+- `restart`: Restarts the node under tests, closes the existing connection, cleans up the existing test files, and restarts bitcoind. Note that it's useful to print a `RESTART` log when verbose logging is activated, e.g.
+
+        if self.config.getoption('verbose'):
+            print("[RESTART]")
+
+- `connect`: Create a connection to the node under test using the provided `connprivkey`.
+- `getblockheight`: Return the blockcount from bitcoind
+- `trim_blocks`: invalidate bitcoind blocks until `newheight`
+- `add_blocks`: Send provided `txs` (if any). Generate `n` new blocks.
```

Comment:
Why does this need to be done by child runners? It feels like it should be the same for all runners, isn't it?

Reply:
Ah it's because each node could use a different bitcoin backend, right?
So nodes should also package their bitcoin backend of choice (`bitcoind` in most cases)?
Maybe it's worth providing one implementation for `bitcoind` that all runners who use `bitcoind` can share.

Reply:
>Ah it's because each node could use a different bitcoin backend, right?

I'm thinking that the interface of the Bitcoin class can be unique for all the implementation, at least a parent of all implementation.

If I'm not missing nothings (and I can), these actions are required by the lightning protocol (not trim block), and all the implementation of backend should have them, right? 

A good idea could be to have a generic Backend implementation and with that implement the bitcoin backend. With these solutions, we can cover all the possibilities.

Reply:
> Maybe it's worth providing one implementation for bitcoind that all runners who use bitcoind can share.

It should be just a matter of pulling the `Bitcoind` class out of clightning.py and moving it somewhere that's accessible?

See:
https://github.com/rustyrussell/lnprototest/blob/master/lnprototest/clightning/clightning.py#L28-L73